File: 贪心算法/easy/860.柠檬水找零.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=860 lang=cpp
 *
 * [860] 柠檬水找零
 */

// @lc code=start
class Solution
{
public:
    bool lemonadeChange(vector<int> &bills)
    {
        int five = 0;
        int ten = 0;
        int twenty = 0;

        for (int i = 0; i < bills.size(); i++)
        {
            // 5块直接收，记录5块的张数
            if (bills[i] == 5)
            {
                five++;
                continue;
            }

            // 10块直接找回5块
            if (bills[i] == 10)
            {
                // 已经没有5元可以找回
                if (five <= 0)
                {
                    return false;
                }
                ten++;
                five--;
            }

            // 收到20块钱
            if (bills[i] == 20)
            {
                // 找回一个10块和5块，因为10块只能找回给20,5块可以找回给20和10块
                // 优先使用10块
                if (ten > 0 && five > 0)
                {
                    ten--;
                    five--;
                    twenty++;
                }
                // 找回3个5块
                else if (five >= 3)
                {
                    five -= 3;
                    twenty++;
                }
                else
                {
                    return false;
                }
            }
        }

        return true;
    }
};
// @lc code=end

File: 贪心算法/easy/860.柠檬水找零.ts
/*
 * @lc app=leetcode.cn id=860 lang=typescript
 *
 * [860] 柠檬水找零
 */

// @lc code=start
function lemonadeChange(bills: number[]): boolean {
  let priceMap: any = { 5: 0, 10: 0 }

  for (let i = 0; i < bills.length; i++) {

    switch (bills[i]) {
      case 5:
        // 收到5块不用找零
        priceMap[5] = priceMap[5] + 1
        break
      case 10:
        // 收到10块，没有5块找零，则找不开
        if (priceMap[5] === 0) {
          return false
        }
        priceMap[5] = priceMap[5] - 1
        priceMap[10] = priceMap[10] + 1
        break
      case 20:
        // 收到20，没有10块，且5块小于3张，找不开
        if (priceMap[10] <= 0) {
          if (priceMap[5] < 3) {
            return false
          }
          priceMap[5] = priceMap[5] - 3
        } else {
          // 收到20，有10块，5块小于1张，找不开
          if (priceMap[5] < 1) {
            return false
          }
          priceMap[10] = priceMap[10] - 1
          priceMap[5] = priceMap[5] - 1
        }

        break
    }
  }

  return true
};
// @lc code=end

